Feature/RcnnPose

yskim041 opened the pull request:
A new perception class for R-CNN + Moped.

***

**Before creating a pull request**

- [x] Document new methods and classes
- [x] Format code with `make format`

**Before merging a pull request**

- [x] Set version target by selecting a milestone on the right side
- [x] Summarize this change in `CHANGELOG.md`

brianhou commented on src/perception/RcnnMopedModule.cpp:
```diff
@@ -0,0 +1,163 @@
+
+
+//=============================================================================
+const dart::dynamics::SkeletonPtr makeBodyFromURDF(const std::string& uri,
```
We have this helper function in a few places already, so perhaps we should create `aikido::io::util` and add this method there. When we do that, we should add a `resourceRetriever` argument (so we don't have the overhead of initializing the `CatkinResourceRetriever` multiple times) like here: https://github.com/personalrobotics/herb3_demos/blob/master/soda_handoff/src/soda_handoff.cpp#L36

yskim041 replied:
Removed this part. I used dart::utils::DartLoader instead.

brianhou commented on src/perception/RcnnMopedModule.cpp:
```diff
@@ -0,0 +1,163 @@
+    Eigen::Isometry3d link_offset = mReferenceLink->getWorldTransform();
+    skel_pose = link_offset * skel_pose;
+
+    std::string skel_name = marker_transform.text;
```
I think this should be in a YAML database or something.

brianhou commented on src/perception/RcnnMopedModule.cpp:
```diff
@@ -0,0 +1,163 @@
+        makeBodyFromURDF(skel_uri, skel_pose);
+    objSkeletonPtr->setName(skel_name);
+
+    dart::dynamics::SkeletonPtr env_skeleton = env->getSkeleton(skel_name);
```
If there are multiple objects with the same `skel_name`, how do we disambiguate between them? Or does MOPED-lite ensure that each name is unique? How does it know that one Coke can is `can_coke(1)` vs `can_coke(2)`?

yskim041 replied:
I made this part relies on `obj_id` provided from `RcnnPose` module. I will make `RcnnPose` handling all naming conflicts so that this receiver does not need to take care of the object detection.

brianhou commented on src/perception/RcnnMopedModule.cpp:
```diff
@@ -0,0 +1,163 @@
+
+    if (env_skeleton != nullptr)
+    {
+      env->removeSkeleton(env_skeleton);
```
This might be related to my question above, but why are we removing this skeleton? It seems that these skeletons would be from a previous call to `detectObjects`, but what if MOPED-lite detected fewer cans than the last call? It seems like the extra cans wouldn't be removed.

yskim041 replied:
I also think that it will be better if the perception module decides everything including removing a disappeared object. Now this `RcnnMopedModule` will remove a skeleton with `obj_id` only if it gets "None" in `obj_key`.

brianhou commented on include/aikido/perception/RcnnMopedModule.hpp:
```diff
@@ -0,0 +1,50 @@
+      ros::NodeHandle nodeHandle,
+      std::string markerTopic,
+      dart::common::ResourceRetrieverPtr resourceRetriever,
+      std::string canUri,
```
Actually, it seems that this parameter / corresponding field aren't used at all. We should remove this.

yskim041 replied:
I added `ObjectDatabase` which is similar to `YamlAprilTagsDatabase`. I generalized this part by using `ObjectDatabase`, and made a corresponding JSON file in personalrobotics/pr-ordata#28 .

brianhou commented on src/perception/RcnnMopedModule.cpp:
```diff
@@ -0,0 +1,163 @@
+{
+  bool any_valid = false;
+  
+  tf::TransformListener listener(mNodeHandle);
```
Is this used?

yskim041 replied:
Removed this TransformListener.

yskim041 commented on include/aikido/perception/RcnnMopedModule.hpp:
```diff
@@ -0,0 +1,50 @@
+      ros::Duration timeout = ros::Duration(0.0),
+      ros::Time timestamp = ros::Time(0.0)) override;
+
+private:
```
added documentation

yskim041 commented on include/aikido/perception/RcnnMopedModule.hpp:
```diff
@@ -0,0 +1,50 @@
+class RcnnMopedModule : public PerceptionModule
+{
+public:
+  RcnnMopedModule(
```
added documentation

yskim041 commented:
I addressed all issues. I want to generalize this module as a common receiver for any kinds of rcnn+<pose detection> algorithms I will implement separately.
After getting approvals from reviewers, I will change the file names from `RcnnMoped*` to `RcnnPose*`.

brianhou commented on include/aikido/perception/ObjectDatabase.hpp:
```diff
@@ -0,0 +1,57 @@
+
+  virtual ~ObjectDatabase() = default;
+
+  bool getObjectByKey(
```
Can you document this? In particular, it seems from the code that it should have `\param[out] obj_name` and `\param[out] obj_resource`.

brianhou commented on include/aikido/perception/RcnnMopedModule.hpp:
```diff
@@ -0,0 +1,80 @@
+/// It receives input from the RcnnPose running in a separate process,
+/// by subscribing to a visualization_msg::MarkerArray ROS topic published by
+/// the RcnnPose node. It uses a \c ObjectDatabase to resolve each marker to a
+/// \c dart::common::Uri, and updates the environment which is a list of \c
```
"updates the environment which is an `aikido::planner::World`"

brianhou commented on include/aikido/perception/RcnnMopedModule.hpp:
```diff
@@ -0,0 +1,80 @@
+  RcnnPoseModule(
+      ros::NodeHandle nodeHandle,
+      std::string markerTopic,
+      std::shared_ptr<ObjectDatabase> configData,
```
This parameter is missing documentation.

yskim041 replied:
added documentation for `configData`

brianhou commented on src/perception/ObjectDatabase.cpp:
```diff
@@ -0,0 +1,85 @@
+  // Read the JSON file into string
+  if (!resourceRetriever)
+  {
+    throw std::invalid_argument("ResourceRetrieverPtr given is null!");
```
I think `ResourceRetriever is null` matches the rest of the error messages in `aikido` better.

brianhou commented on src/perception/ObjectDatabase.cpp:
```diff
@@ -0,0 +1,85 @@
+    throw std::invalid_argument("ResourceRetrieverPtr given is null!");
+  }
+
+  const dart::common::ResourcePtr resource
```
I think this and the following section that puts the file in a string can be replaced by `resourceRetriever->readAll(configDataURI)`.

brianhou commented on src/perception/ObjectDatabase.cpp:
```diff
@@ -0,0 +1,85 @@
+  mObjData = YAML::Load(content);
+}
+
+//==============================================================================
```
It seems that this function will either throw an error or succeed. I think rather than returning `bool` (which isn't very helpful if we're throwing errors anyway), we can just make the return type `void`.

yskim041 replied:
I agree with this point. I changed the return type to `void`, and removed all returns.

brianhou commented on src/perception/ObjectDatabase.cpp:
```diff
@@ -0,0 +1,85 @@
+{
+  // Get name of object and pose for a given tag ID
+  YAML::Node obj_node = mObjData[_obj_key];
+  if (obj_node)
```
I think we can restructure this as `if (!obj_node) ...` then just dedenting the block that is currently inside the `if`.

brianhou commented on src/perception/RcnnMopedModule.cpp:
```diff
@@ -0,0 +1,199 @@
+#include <aikido/io/CatkinResourceRetriever.hpp>
+#include <aikido/perception/shape_conversions.hpp>
+
+using dart::dynamics::SkeletonPtr;
```
It doesn't seem like this is being used.

brianhou commented on src/perception/RcnnMopedModule.cpp:
```diff
@@ -0,0 +1,199 @@
+{
+  // Checks detected objects, looks up the database,
+  // and adds new skeletons to the world env
+  bool any_valid = false;
```
Maybe `any_detected` would be a better name than `any_valid`.

brianhou commented on src/perception/RcnnMopedModule.cpp:
```diff
@@ -0,0 +1,199 @@
+      continue;
+    }
+
+    // If obj_key is "None",
```
When would `obj_key` be None? Does the RCNN keep track of that somehow?

yskim041 replied:
`RcnnPose` will set `obj_key` None when an object disappears on its list.

brianhou commented on src/perception/RcnnMopedModule.cpp:
```diff
@@ -0,0 +1,199 @@
+    bool is_new_obj;
+    dart::dynamics::SkeletonPtr obj_skeleton;
+
+    // Check if skel in skel_list
```
`skel_list` -> `World`
